File: frontend/services/billingService.ts
import { apiClient } from '@/lib/api';

export interface BillingItem {
  id: string;
  serviceName: string;
  qty: number;
  rate: number;
  amount: number;
}

export interface BillingRecord {
  id: string;
  billNumber: string;
  patientId: number;
  patientName: string;
  patientUhid: string;
  appointmentId?: string | null;
  visitType: string;
  doctorId: string;
  doctorName: string;
  date: string;
  subtotal: number;
  discount: number;
  tax: number;
  netAmount: number;
  paidAmount: number;
  balanceAmount: number;
  status: 'Paid' | 'Pending' | 'Partial';
  paymentMode: string;
  transactionId?: string | null;
  createdAt: string;
  items: BillingItem[];
}

export interface BillingListParams {
  page?: number;
  pageSize?: number;
  search?: string;
  status?: string;
  doctorId?: string;
  fromDate?: string;
  toDate?: string;
}

export interface PagedBillingResult {
  items: BillingRecord[];
  page: number;
  pageSize: number;
  totalRecords: number;
  totalPages: number;
}

const toNumber = (value: unknown, fallback: number): number => {
  const parsed = Number(value);
  return Number.isFinite(parsed) ? parsed : fallback;
};

const normalizeBillingRecord = (raw: Record<string, unknown>): BillingRecord => {
  const paidAmount = toNumber(raw.paidAmount ?? raw.PaidAmount, 0);
  const netAmount = toNumber(raw.netAmount ?? raw.NetAmount, 0);

  let status = String(raw.status ?? raw.Status ?? '').trim();
  if (!status) {
    status = paidAmount <= 0 ? 'Pending' : paidAmount >= netAmount ? 'Paid' : 'Partial';
  }

  const rawItems = raw.items ?? raw.Items ?? [];
  const items: BillingItem[] = Array.isArray(rawItems)
    ? rawItems.map((item) => {
        const row = (item ?? {}) as Record<string, unknown>;
        return {
          id: String(row.id ?? row.Id ?? ''),
          serviceName: String(row.serviceName ?? row.ServiceName ?? ''),
          qty: toNumber(row.qty ?? row.Qty, 0),
          rate: toNumber(row.rate ?? row.Rate, 0),
          amount: toNumber(row.amount ?? row.Amount, 0),
        };
      })
    : [];

  return {
    id: String(raw.id ?? raw.Id ?? ''),
    billNumber: String(raw.billNumber ?? raw.BillNumber ?? ''),
    patientId: toNumber(raw.patientId ?? raw.PatientId, 0),
    patientName: String(raw.patientName ?? raw.PatientName ?? ''),
    patientUhid: String(raw.patientUhid ?? raw.PatientUhid ?? ''),
    appointmentId: (raw.appointmentId ?? raw.AppointmentId ?? null) as string | null,
    visitType: String(raw.visitType ?? raw.VisitType ?? ''),
    doctorId: String(raw.doctorId ?? raw.DoctorId ?? ''),
    doctorName: String(raw.doctorName ?? raw.DoctorName ?? ''),
    date: String(raw.date ?? raw.Date ?? ''),
    subtotal: toNumber(raw.subtotal ?? raw.Subtotal, 0),
    discount: toNumber(raw.discount ?? raw.Discount, 0),
    tax: toNumber(raw.tax ?? raw.Tax, 0),
    netAmount,
    paidAmount,
    balanceAmount: toNumber(raw.balanceAmount ?? raw.BalanceAmount, Math.max(0, netAmount - paidAmount)),
    status: status as BillingRecord['status'],
    paymentMode: String(raw.paymentMode ?? raw.PaymentMode ?? ''),
    transactionId: (raw.transactionId ?? raw.TransactionId ?? null) as string | null,
    createdAt: String(raw.createdAt ?? raw.CreatedAt ?? ''),
    items,
  };
};

const normalizePagedBilling = (payload: unknown, params: BillingListParams): PagedBillingResult => {
  if (Array.isArray(payload)) {
    const items = payload.map((item) => normalizeBillingRecord((item ?? {}) as Record<string, unknown>));
    const page = params.page ?? 1;
    const pageSize = params.pageSize ?? 10;
    const totalRecords = items.length;
    return {
      items,
      page,
      pageSize,
      totalRecords,
      totalPages: totalRecords === 0 ? 0 : Math.ceil(totalRecords / pageSize),
    };
  }

  const data = (payload ?? {}) as Record<string, unknown>;
  const rawItems = data.items ?? data.Items ?? [];
  const items = Array.isArray(rawItems)
    ? rawItems.map((item) => normalizeBillingRecord((item ?? {}) as Record<string, unknown>))
    : [];

  const page = toNumber(data.page ?? data.Page, params.page ?? 1);
  const pageSize = toNumber(data.pageSize ?? data.PageSize, params.pageSize ?? 10);
  const totalRecords = toNumber(data.totalRecords ?? data.TotalRecords, items.length);
  const totalPages = toNumber(
    data.totalPages ?? data.TotalPages,
    totalRecords === 0 ? 0 : Math.ceil(totalRecords / pageSize)
  );

  return {
    items,
    page,
    pageSize,
    totalRecords,
    totalPages,
  };
};

class BillingService {
  async getBillings(params: BillingListParams = {}): Promise<PagedBillingResult> {
    const response = await apiClient.get('/api/billing', { params });
    return normalizePagedBilling(response.data?.data, params);
  }

  async getAll(): Promise<BillingRecord[]> {
    const result = await this.getBillings({ page: 1, pageSize: 100 });
    return result.items;
  }

  async getById(id: string): Promise<BillingRecord> {
    const response = await apiClient.get(`/api/billing/${id}`);
    return normalizeBillingRecord((response.data?.data ?? {}) as Record<string, unknown>);
  }

  async getByBillNumber(billNumber: string): Promise<BillingRecord> {
    const response = await apiClient.get(`/api/billing/by-number/${billNumber}`);
    return normalizeBillingRecord((response.data?.data ?? {}) as Record<string, unknown>);
  }

  async create(data: Record<string, unknown>): Promise<BillingRecord> {
    const response = await apiClient.post('/api/billing', data);
    return normalizeBillingRecord((response.data?.data ?? {}) as Record<string, unknown>);
  }

  async update(id: string, data: Record<string, unknown>): Promise<BillingRecord> {
    const response = await apiClient.put(`/api/billing/${id}`, data);
    return normalizeBillingRecord((response.data?.data ?? {}) as Record<string, unknown>);
  }

  async cancel(id: string): Promise<void> {
    await apiClient.delete(`/api/billing/${id}/cancel`);
  }
}

export default new BillingService();

File: backend/src/HMS.Application/Features/Billing/BillingService.cs
using HMS.Application.Repositories;
using HMS.Domain.Entities;
using DomainBilling = HMS.Domain.Entities.Billing;

namespace HMS.Application.Features.Billing;

public class BillingService : IBillingService
{
    private static readonly HashSet<string> AllowedVisitTypes = new(StringComparer.OrdinalIgnoreCase)
    {
        "consultation",
        "follow-up",
        "procedure",
    };

    private static readonly HashSet<string> AllowedPaymentModes = new(StringComparer.OrdinalIgnoreCase)
    {
        "cash",
        "upi",
        "card",
    };

    private readonly IBillingRepository _billingRepository;
    private readonly IPatientRepository _patientRepository;
    private readonly IAppointmentRepository _appointmentRepository;
    private readonly IDoctorRepository _doctorRepository;

    public BillingService(
        IBillingRepository billingRepository,
        IPatientRepository patientRepository,
        IAppointmentRepository appointmentRepository,
        IDoctorRepository doctorRepository)
    {
        _billingRepository = billingRepository;
        _patientRepository = patientRepository;
        _appointmentRepository = appointmentRepository;
        _doctorRepository = doctorRepository;
    }

    public async Task<List<BillingDto>> GetAllAsync()
    {
        var billings = await _billingRepository.GetAllAsync();
        return billings.Select(MapToDto).ToList();
    }

    public async Task<BillingPagedResultDto> GetPagedAsync(BillingListQueryDto query)
    {
        var page = Math.Max(1, query.Page);
        var pageSize = Math.Clamp(query.PageSize, 1, 100);

        var (items, totalRecords) = await _billingRepository.GetPagedAsync(
            query.Search,
            query.Status,
            query.DoctorId,
            query.FromDate,
            query.ToDate,
            page,
            pageSize);

        return new BillingPagedResultDto
        {
            Items = items.Select(MapToDto).ToList(),
            Page = page,
            PageSize = pageSize,
            TotalRecords = totalRecords,
            TotalPages = totalRecords == 0 ? 0 : (int)Math.Ceiling(totalRecords / (double)pageSize),
        };
    }

    public async Task<BillingDto?> GetByIdAsync(string id)
    {
        var billing = await _billingRepository.GetByIdAsync(id);
        return billing == null ? null : MapToDto(billing);
    }

    public async Task<BillingDto?> GetByBillNumberAsync(string billNumber)
    {
        var billing = await _billingRepository.GetByBillNumberAsync(billNumber);
        return billing == null ? null : MapToDto(billing);
    }

    public async Task<BillingDto> CreateAsync(CreateBillingDto request)
    {
        await ValidateRequestAsync(request);

        var sanitizedItems = request.Items
            .Where(item => !string.IsNullOrWhiteSpace(item.ServiceName))
            .Select(item =>
            {
                var qty = Math.Max(1, item.Qty);
                var rate = Math.Max(0m, item.Rate);
                return new CreateBillingItemDto
                {
                    ServiceName = item.ServiceName.Trim(),
                    Qty = qty,
                    Rate = rate,
                };
            })
            .ToList();

        if (sanitizedItems.Count == 0)
        {
            throw new ArgumentException("At least one billing item is required");
        }

        var subtotal = sanitizedItems.Sum(item => item.Qty * item.Rate);
        var discount = CalculateDiscount(subtotal, request.DiscountValue, request.DiscountType);
        var tax = Math.Max(0m, request.Tax);
        var netAmount = subtotal - discount + tax;

        if (netAmount <= 0)
        {
            throw new ArgumentException("Net amount must be greater than 0");
        }

        var paidAmount = Math.Max(0m, request.PaidAmount);

        var billing = new DomainBilling
        {
            BillNumber = await _billingRepository.GetNextBillNumberAsync(),
            PatientId = request.PatientId,
            AppointmentId = string.IsNullOrWhiteSpace(request.AppointmentId) ? null : request.AppointmentId.Trim(),
            VisitType = request.VisitType.Trim().ToLowerInvariant(),
            DoctorId = request.DoctorId.Trim(),
            Date = request.Date == default ? DateTime.UtcNow.Date : request.Date.Date,
            Subtotal = subtotal,
            Discount = discount,
            Tax = tax,
            NetAmount = netAmount,
            PaidAmount = paidAmount,
            PaymentMode = request.PaymentMode.Trim().ToLowerInvariant(),
            TransactionId = string.IsNullOrWhiteSpace(request.TransactionId) ? null : request.TransactionId.Trim(),
            Items = sanitizedItems.Select(item => new BillingItem
            {
                ServiceName = item.ServiceName,
                Qty = item.Qty,
                Rate = item.Rate,
                Amount = item.Qty * item.Rate,
            }).ToList(),
        };

        var created = await _billingRepository.CreateAsync(billing);
        return MapToDto(created);
    }

    public async Task<BillingDto> UpdateAsync(string id, UpdateBillingDto request)
    {
        if (string.IsNullOrWhiteSpace(id))
        {
            throw new ArgumentException("Billing ID is required", nameof(id));
        }

        var existing = await _billingRepository.GetByIdAsync(id.Trim())
            ?? throw new KeyNotFoundException($"Billing with ID {id} not found");

        if (existing.CreatedAt.Date != DateTime.UtcNow.Date)
        {
            throw new InvalidOperationException("Editing is allowed only for records created today");
        }

        await ValidateRequestAsync(request);

        var sanitizedItems = request.Items
            .Where(item => !string.IsNullOrWhiteSpace(item.ServiceName))
            .Select(item =>
            {
                var qty = Math.Max(1, item.Qty);
                var rate = Math.Max(0m, item.Rate);
                return new CreateBillingItemDto
                {
                    ServiceName = item.ServiceName.Trim(),
                    Qty = qty,
                    Rate = rate,
                };
            })
            .ToList();

        if (sanitizedItems.Count == 0)
        {
            throw new ArgumentException("At least one billing item is required");
        }

        var subtotal = sanitizedItems.Sum(item => item.Qty * item.Rate);
        var discount = CalculateDiscount(subtotal, request.DiscountValue, request.DiscountType);
        var tax = Math.Max(0m, request.Tax);
        var netAmount = subtotal - discount + tax;

        if (netAmount <= 0)
        {
            throw new ArgumentException("Net amount must be greater than 0");
        }

        var paidAmount = Math.Max(0m, request.PaidAmount);
        if (paidAmount > netAmount)
        {
            throw new ArgumentException("Paid amount cannot be greater than net amount", nameof(request.PaidAmount));
        }

        existing.PatientId = request.PatientId;
        existing.AppointmentId = string.IsNullOrWhiteSpace(request.AppointmentId) ? null : request.AppointmentId.Trim();
        existing.VisitType = request.VisitType.Trim().ToLowerInvariant();
        existing.DoctorId = request.DoctorId.Trim();
        existing.Date = request.Date == default ? DateTime.UtcNow.Date : request.Date.Date;
        existing.Subtotal = subtotal;
        existing.Discount = discount;
        existing.Tax = tax;
        existing.NetAmount = netAmount;
        existing.PaidAmount = paidAmount;
        existing.PaymentMode = request.PaymentMode.Trim().ToLowerInvariant();
        existing.TransactionId = string.IsNullOrWhiteSpace(request.TransactionId) ? null : request.TransactionId.Trim();

        existing.Items.Clear();
        foreach (var item in sanitizedItems)
        {
            existing.Items.Add(new BillingItem
            {
                ServiceName = item.ServiceName,
                Qty = item.Qty,
                Rate = item.Rate,
                Amount = item.Qty * item.Rate,
            });
        }

        var updated = await _billingRepository.UpdateAsync(existing);
        return MapToDto(updated);
    }

    public async Task CancelAsync(string id)
    {
        if (string.IsNullOrWhiteSpace(id))
        {
            throw new ArgumentException("Billing ID is required", nameof(id));
        }

        await _billingRepository.CancelAsync(id.Trim());
    }

    private async Task ValidateRequestAsync(CreateBillingDto request)
    {
        if (request.PatientId <= 0)
        {
            throw new ArgumentException("Patient selection is required", nameof(request.PatientId));
        }

        if (string.IsNullOrWhiteSpace(request.VisitType) || !AllowedVisitTypes.Contains(request.VisitType.Trim()))
        {
            throw new ArgumentException("Visit type must be one of: consultation, follow-up, procedure", nameof(request.VisitType));
        }

        if (string.IsNullOrWhiteSpace(request.DoctorId))
        {
            throw new ArgumentException("Doctor selection is required", nameof(request.DoctorId));
        }

        if (request.Items == null || request.Items.Count == 0)
        {
            throw new ArgumentException("At least one billing item is required", nameof(request.Items));
        }

        if (string.IsNullOrWhiteSpace(request.PaymentMode) || !AllowedPaymentModes.Contains(request.PaymentMode.Trim()))
        {
            throw new ArgumentException("Payment mode must be one of: cash, upi, card", nameof(request.PaymentMode));
        }

        if (!string.Equals(request.PaymentMode.Trim(), "cash", StringComparison.OrdinalIgnoreCase)
            && string.IsNullOrWhiteSpace(request.TransactionId))
        {
            throw new ArgumentException("Transaction ID is required for UPI/Card payments", nameof(request.TransactionId));
        }

        var patient = await _patientRepository.GetByIdAsync(request.PatientId);
        if (patient == null)
        {
            throw new KeyNotFoundException($"Patient with ID {request.PatientId} not found");
        }

        var doctor = await _doctorRepository.GetByIdAsync(request.DoctorId.Trim());
        if (doctor == null)
        {
            throw new KeyNotFoundException($"Doctor with ID {request.DoctorId} not found");
        }

        if (!string.IsNullOrWhiteSpace(request.AppointmentId))
        {
            var appointment = await _appointmentRepository.GetByIdAsync(request.AppointmentId.Trim())
                ?? throw new KeyNotFoundException($"Appointment with ID {request.AppointmentId} not found");

            if (appointment.PatientId != request.PatientId)
            {
                throw new ArgumentException("Selected appointment does not belong to the selected patient");
            }
        }
    }

    private static decimal CalculateDiscount(decimal subtotal, decimal discountValue, string discountType)
    {
        var normalizedType = string.IsNullOrWhiteSpace(discountType)
            ? "amount"
            : discountType.Trim().ToLowerInvariant();

        var value = Math.Max(0m, discountValue);

        if (subtotal <= 0m || value <= 0m)
        {
            return 0m;
        }

        if (normalizedType == "percentage")
        {
            var boundedPercent = Math.Min(100m, value);
            return Math.Round(subtotal * boundedPercent / 100m, 2);
        }

        return Math.Min(subtotal, value);
    }

    private static BillingDto MapToDto(DomainBilling billing)
    {
        return new BillingDto
        {
            Id = billing.Id,
            BillNumber = billing.BillNumber,
            PatientId = billing.PatientId,
            PatientName = billing.Patient?.PatientName ?? string.Empty,
            PatientUhid = billing.Patient?.Uhid ?? string.Empty,
            AppointmentId = billing.AppointmentId,
            VisitType = billing.VisitType,
            DoctorId = billing.DoctorId,
            DoctorName = billing.Doctor?.Name ?? string.Empty,
            Date = billing.Date,
            Subtotal = billing.Subtotal,
            Discount = billing.Discount,
            Tax = billing.Tax,
            NetAmount = billing.NetAmount,
            PaidAmount = billing.PaidAmount,
            BalanceAmount = Math.Max(0m, billing.NetAmount - billing.PaidAmount),
            Status = GetPaymentStatus(billing.NetAmount, billing.PaidAmount),
            PaymentMode = billing.PaymentMode,
            TransactionId = billing.TransactionId,
            CreatedAt = billing.CreatedAt,
            Items = billing.Items.Select(item => new BillingItemDto
            {
                Id = item.Id,
                ServiceName = item.ServiceName,
                Qty = item.Qty,
                Rate = item.Rate,
                Amount = item.Amount,
            }).ToList(),
        };
    }

    private static string GetPaymentStatus(decimal netAmount, decimal paidAmount)
    {
        if (paidAmount <= 0m)
        {
            return "Pending";
        }

        if (paidAmount >= netAmount)
        {
            return "Paid";
        }

        return "Partial";
    }
}
